include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -6,425 +6,762 @@
+
+template<typename UInt64> NBL_PARTIAL_REQ_TOP(is_same_v<UInt64, uint64_t>)
+struct find_msb_helper<UInt64 NBL_PARTIAL_REQ_BOT(is_same_v<UInt64, uint64_t>) >
+{
+	using return_t = int32_t;
+	static return_t __call(NBL_CONST_REF_ARG(UInt64) val)
+	{
+		const uint32_t highBits = uint32_t(val >> 32);
+		const int32_t highMsb = find_msb_helper<uint32_t>::__call(highBits);
 
-DEFINE_BUILTIN_VECTOR_SPECIALIZATION(float16_t, BUILTIN_VECTOR_SPECIALIZATION_RET_VAL)
-DEFINE_BUILTIN_VECTOR_SPECIALIZATION(float32_t, BUILTIN_VECTOR_SPECIALIZATION_RET_VAL)
-DEFINE_BUILTIN_VECTOR_SPECIALIZATION(float64_t, BUILTIN_VECTOR_SPECIALIZATION_RET_VAL)
+		if (highMsb == -1)
+		{
+			const uint32_t lowBits = uint32_t(val);
+			const int32_t lowMsb = find_msb_helper<uint32_t>::__call(lowBits);
+			if (lowMsb == -1)
+				return -1;
 
-#undef BUILTIN_VECTOR_SPECIALIZATION_RET_VAL
-#undef DEFINE_BUILTIN_VECTOR_SPECIALIZATION
+			return lowMsb;
+		}
 
-template<typename Integer>
-struct find_msb_helper;
+		return highMsb + 32;
+	}
+};
+template<typename UInt64> NBL_PARTIAL_REQ_TOP(is_same_v<UInt64, uint64_t>)
+struct find_lsb_helper<UInt64 NBL_PARTIAL_REQ_BOT(is_same_v<UInt64, uint64_t>) >
+{
+	static int32_t __call(NBL_CONST_REF_ARG(uint64_t) val)
+	{
+		const uint32_t lowBits = uint32_t(val);
+		const int32_t lowLsb = find_lsb_helper<uint32_t>::__call(lowBits);
 
-template<>
-struct find_msb_helper<uint32_t>
+		if (lowLsb == -1)
+		{
+			const uint32_t highBits = uint32_t(val >> 32);
+			const int32_t highLsb = find_lsb_helper<uint32_t>::__call(highBits);
+			if (highLsb == -1)
+				return -1;
+			else
+				return 32 + highLsb;
+		}
+
+		return lowLsb;
+	}
+};
+
+template<typename SquareMatrix>
+NBL_PARTIAL_REQ_TOP(concepts::Matrix<SquareMatrix>&& matrix_traits<SquareMatrix>::Square)
+struct inverse_helper<SquareMatrix NBL_PARTIAL_REQ_BOT(concepts::Matrix<SquareMatrix>&& matrix_traits<SquareMatrix>::Square) >
 {
-	static int32_t findMSB(NBL_CONST_REF_ARG(uint32_t) val)
+	static SquareMatrix __call(NBL_CONST_REF_ARG(SquareMatrix) mat)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findUMsb(val);
-#else
-		return glm::findMSB(val);
-#endif
+		return spirv::matrixInverse(mat);
 	}
 };
 
-template<>
-struct find_msb_helper<int32_t>
+template<typename T, typename U> NBL_PARTIAL_REQ_TOP(always_true<decltype(spirv::fMix<T>(experimental::declval<T>(), experimental::declval<T>(), experimental::declval<U>()))>)
+struct mix_helper<T, U NBL_PARTIAL_REQ_BOT(always_true<decltype(spirv::fMix<T>(experimental::declval<T>(), experimental::declval<T>(), experimental::declval<U>()))>) >
 {
-	static int32_t findMSB(NBL_CONST_REF_ARG(int32_t) val)
+	using return_t = conditional_t<is_vector_v<T>, vector<typename vector_traits<T>::scalar_type, vector_traits<T>::Dimension>, T>;
+	static inline return_t __call(const T x, const T y, const U a)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findSMsb(val);
-#else
-		return glm::findMSB(val);
-#endif
+		T aAsT = a;
+		return spirv::fMix<T>(x, y, aAsT);
```

Comment:
something is off here, you're testing if `fMix` can be called with T,T,U but fmix can only be called with T,T,T ?

Reply:
> is there no SPIR-V intrinsic for GLSL's `mix(T,T,bool)` ?

yes
![image](https://github.com/user-attachments/assets/3a5886fc-b6fc-463a-a067-01c1bf6cfbec)

Reply:
ok I see, then a common C++ && HLSL partial spec needs to be made for `U!=T`